Adding clear errors to RNN's and FFN's when the input shape is wrong

NippunSharma opened the pull request:
This PR is in reference to #2071 opened by @rcurtin . 
The issue was to add more clear errors whenever there was some wrong input shape in the first layer of FFN / RNN.

I added an `InShapeVisitor()` which checks for the expected input shape for each layer (returned by the function `InputShape()`).
For many layers the `InputShape()` function should return the same value as the `InputSize()` (for example `linear`) but for some other layers such as `convolution`, `transposed convolution` the `InputSize()` method just returns `inSize` (the number of input maps) and not the total input shape (which would be `inputHeight*inputWidth*inSize`) which might be needed in this case. I have added the `InputShape()` method to `linear` layer (I will add to the rest of the layers after I know that I am going in the correct direction). Also, `CheckInputShape` method has been added inside `FFN` (will add to `RNN`) that checks the input shape whenever `Train`, `Predict` or a specific overload of `Evaluate`(that has `predictors` as a parameter) is called.                             

Things yet to be added in this PR:

- [x] `InputShape()` (already added to linear, convolution, lstm, fast_lstm)
   -> Have to handle `recurrent.hpp` (done)
   -> Add to `NoisyLinear` (done)
   -> Add to `LinearNoBias` (done)
   -> Add to `Atrous Conv` (done)
   -> Add to `Transposed Conv` (done)
   -> Add to `GRU` (done)
   -> Add to `Bilinear Interpolation` (done)
   -> Add to `Glimpse` (done)
   -> Add to `Highway` (done)
   -> Add to `Linear3D` (done)
   -> Add to `Linear Norm` (done)
   -> Add to `Mini Batch Discrimination` (done)
   -> Add to `Radial Basis Function` (done)
  -> Add to `Reparametrization` (done)
  -> Add to `Multihead Attention` (done)
  -> Add to `Sequential` (done)
  -> Add to `Positional Encoding` (done)

- [x] `CheckInputShape` method to `RNN`.

himanshupathak21061998 commented on src/mlpack/methods/ann/ffn_impl.hpp:
```diff
@@ -50,6 +51,33 @@ FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::~FFN()
 
+template<typename OutputLayerType, typename InitializationRuleType,
+         typename... CustomLayers>
+void FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::CheckInputShape(
```
Hey @NippunSharma thanks for contributing.
I think if you add this in `util` it will be more better. What do you think? @zoq

zoq replied:
You mean an `util` folder in the ann folder, I think that would make sense.

NippunSharma replied:
@zoq , @himanshupathak21061998 I think this would be a good idea. I will add the `util` folder inside `ann` and put the `CheckInputShape` function there.

zoq commented on src/mlpack/methods/ann/ffn.hpp:
```diff
@@ -324,6 +325,14 @@ class FFN
   arma::mat& Predictors() { return predictors; }
 
+  /**
+   * Check wether the input shape is consistent with the layer requirements.
```
```suggestion
   * Check whether the input shape is consistent with the layer requirements.
```

zoq commented on src/mlpack/methods/ann/ffn.hpp:
```diff
@@ -324,6 +325,14 @@ class FFN
+  /**
+   * Check wether the input shape is consistent with the layer requirements.
+   * 
+   * @param inputShape shape of the input
+   * @param functionName function that checks the input size
```
```suggestion
   * @param inputShape Dhape of the input
   * @param functionName Function that checks the input size
```

zoq commented on src/mlpack/methods/ann/ffn.hpp:
```diff
@@ -324,6 +325,14 @@ class FFN
+   * @param inputShape shape of the input
+   * @param functionName function that checks the input size
+   */  
+  void CheckInputShape(size_t inputShape, std::string functionName);
```
```suggestion
  void CheckInputShape(const size_t inputShape, const std::string& functionName);
```

To avoid an extra copy.

zoq commented on src/mlpack/methods/ann/ffn_impl.hpp:
```diff
@@ -50,6 +51,33 @@ FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::~FFN()
+    }
+    else
+    {
+      std::string estr = "FFN<>::" + functionName + ": "; 
+                  estr += "the first layer of the network expects ";
+                  estr += std::to_string(layerInShape) + " elements, ";
+                  estr += "but the input has " + std::to_string(inputShape) + " rows! ";
+      throw std::logic_error(estr);
```
Not sure, if the function name is necessary, if we apply the same functionality to the GAN and RNN class, I guess it would make it easier if we don't have the function name here.

NippunSharma replied:
I added function name because the user might use both `Train()` and `Predict()` in the same program and function name might help in debugging and informing the name of function where the user gave wrong input.

rcurtin commented on src/mlpack/methods/ann/layer/convolution.hpp:
```diff
@@ -259,6 +259,12 @@ class Convolution
+  //! Get the shape of the input.
+  size_t InputShape() const
+  {
+    return inputHeight*inputWidth*inSize;
```
```suggestion
    return inputHeight * inputWidth * inSize;
```

This is just to match the mlpack style guide.  Do you think you can make this change (horizontal whitespace) wherever it's necessary in this PR?

NippunSharma replied:
Sure, I will make the changes 👍

rcurtin commented on src/mlpack/methods/ann/layer/recurrent_impl.hpp:
```diff
@@ -126,6 +127,36 @@ Recurrent<InputDataType, OutputDataType, CustomLayers...>::Recurrent(
+  size_t inputShapeFeedbackModule = boost::apply_visitor(InShapeVisitor(), feedbackModule);
+  size_t inputShapeTransferModule = boost::apply_visitor(InShapeVisitor(), transferModule);
+
+  if (inputShapeStartModule != 0)
```
```suggestion
  // Return the size of the first module that we have.
  if (inputShapeStartModule != 0)
```

It might be nice to add a comment detailing why we have to have logic like this. :+1:

NippunSharma replied:
Yes, comments will be helpful here, thanks 👍

rcurtin commented on src/mlpack/methods/ann/util/check_input_shape.hpp:
```diff
@@ -0,0 +1,52 @@
+      std::string estr = functionName + ": "; 
+                  estr += "the first layer of the network expects ";
+                  estr += std::to_string(layerInShape) + " elements, ";
+                  estr += "but the input has " + std::to_string(inputShape) + " rows! ";
```
I'd go with `dimensions` instead of `rows` here---someday we may provide bindings from other languages that represent data in row-major form.  For instance, this error won't make much sense if seen in Python, where the number of rows is the number of points, not the number of dimensions like it is in C++.

rcurtin reviewed (commented):
Hey @NippunSharma, this looks great!  I think this strategy will work just fine.  Should we add `InputShape()` to the rest of the layers that we have implemented?  I suppose those that don't have any weights can just omit that method, but I think there are still some layers that would need it. :+1:

NippunSharma commented:
Hi, I have added the `InputShape()` method to more layers (at least the ones that I could understand). If there are any other remaining layers please let me know and I will add to them also. I have also added the `util` folder to `ann` and moved the `CheckInputShape` there.

zoq commented on src/mlpack/methods/ann/layer/recurrent_impl.hpp:
```diff
@@ -126,6 +127,44 @@ Recurrent<InputDataType, OutputDataType, CustomLayers...>::Recurrent(
+         typename... CustomLayers>
+size_t Recurrent<InputDataType, OutputDataType, CustomLayers...>::InputShape() const
+{
+  size_t inputShapeStartModule = boost::apply_visitor(InShapeVisitor(), startModule);
+  size_t inputShapeInputModule = boost::apply_visitor(InShapeVisitor(), inputModule);
+  size_t inputShapeFeedbackModule = boost::apply_visitor(InShapeVisitor(), feedbackModule);
+  size_t inputShapeTransferModule = boost::apply_visitor(InShapeVisitor(), transferModule);
```
I guess the compiler is going to optimize this out anyway, but do you think we could put each of the lines into the corresponding if block? And maybe use `const size_t` to make it obvious that the parameter will not change.

NippunSharma replied:
I have made the changes

NippunSharma commented:
Hi @rcurtin @zoq . Please take a look at this whenever you have time .

zoq commented on src/mlpack/methods/ann/layer/recurrent_impl.hpp:
```diff
@@ -125,6 +126,53 @@ Recurrent<InputDataType, OutputDataType, CustomLayers...>::Recurrent(
+  {
+    return inputShapeStartModule;
+  }
+  // If input shape of first module is 0
```
```suggestion
  // If input shape of first module is 0.
```

Missing stop at the end. There are some other comments below where this applies as well.

zoq reviewed (commented):
Thanks can you add a test to https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/feedforward_network_test.cpp and https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/recurrent_network_test.cpp which checks that we throw an error if the input shape isn't correct?

Here is an example snippet how to check for an exception:

https://github.com/mlpack/mlpack/blob/c5494232bb7333644d59b6db273c2054333fbca9/src/mlpack/tests/ann_layer_test.cpp#L1431-L1435

NippunSharma commented:
@zoq How many tests do you suggest should I add to each ?

NippunSharma commented:
Is there any way using which we can see what error message is thrown ? I think it would be better because an error (due to invalid  matrix multiplication) is thrown anyways when the input shape is wrong.

zoq commented:
> Is there any way using which we can see what error message is thrown ? I think it would be better because an error (due to invalid matrix multiplication) is already thrown when the input shape is wrong.

But that the solution you implemented should be suppress the exception, right?


> @zoq How many tests do you suggest should I add to each ?

One test for the FNN class and one for the RNN class, is fine from my side.

NippunSharma commented:
Okay I will add the tests. 
Yes, the solution suppresses that exception (earlier exception was due to matrix multiplication error which was thrown from within armadillo I think, and was not helpful in debugging) and now a more clear message is thrown (like the one mentioned in #2071 which is handled inside mlpack itself).
My question was that it would be better if we are able to know what error message is there which will tell whether the exception was from mlpack or was from armadillo due to invalid matrix multiplication.

NippunSharma commented:
@zoq I have added the tests. I was able to solve my doubt by using `REQURE_THROWS_MATCHES` instead of `REQUIRE_THROWS_AS` which I read about in `mlpack/src/tests/catch.hpp`. I used `Catch::Matchers::Message` with `REQUIRE_THROWS_MATCHES` which also checks for the correct message to be thrown along with the type of exception.

zoq commented on src/mlpack/tests/feedforward_network_test.cpp:
```diff
@@ -908,3 +908,41 @@ TEST_CASE("OptimizerTest", "[FeedForwardNetworkTest]")
+  model.Add<LogSoftMax<> >();
+
+  std::string expectedMsg = "FFN<>::Train(): ";
+  		 	  expectedMsg += "the first layer of the network expects ";
+  		 	  expectedMsg += std::to_string(trainData.n_rows - 3) + " elements, ";
+  		 	  expectedMsg += "but the input has " + std::to_string(trainData.n_rows) + " dimensions! ";
```
This should align with the line above.

NippunSharma replied:
Yes. somehow my editor was showing this was aligned. changed it now 👍

zoq commented on src/mlpack/tests/feedforward_network_test.cpp:
```diff
@@ -908,3 +908,41 @@ TEST_CASE("OptimizerTest", "[FeedForwardNetworkTest]")
+  ens::DE opt(200, 1000, 0.6, 0.8, 1e-5);
+
+  REQUIRE_THROWS_MATCHES(model.Train(trainData, trainLabels, opt),
+  						 std::logic_error,
+  						 Catch::Matchers::Message(expectedMsg));
```
The second line should be tabbed twice (4 spaces and I guess we can just put `Catch::Matchers::Message(expectedMsg));` on the same line.

NippunSharma replied:
done.

zoq commented on src/mlpack/tests/feedforward_network_test.cpp:
```diff
@@ -908,3 +908,41 @@ TEST_CASE("OptimizerTest", "[FeedForwardNetworkTest]")
+  testData.shed_row(testData.n_rows - 1);
+
+  FFN<NegativeLogLikelihood<>, RandomInitialization, CustomLayer<> > model;
+  // Purposely putting wrong input shape so that error is thrown
```
```suggestion
  // Purposely putting wrong input shape so that error is thrown.
```

Minor style fix, add missing stop at the end.

NippunSharma commented:
Hi @zoq please take a look at this whenever you have time.

zoq commented on src/mlpack/methods/ann/visitor/input_shape_visitor_impl.hpp:
```diff
@@ -0,0 +1,53 @@
+/**
+ * @file input_shape_visitor_impl.hpp
```
```suggestion
 * @file methods/ann/visitor/input_shape_visitor_impl.hpp
```

zoq reviewed (approved):
Thanks for putting this together, I left some last minor comments about the style, but other than that, this looks good from my side.

rcurtin commented on src/mlpack/methods/ann/util/check_input_shape.hpp:
```diff
@@ -0,0 +1,53 @@
+void CheckInputShape(const T& network, const size_t inputShape, 
+                     const std::string& functionName)
+{
+  for (size_t l=0; l<network.size(); ++l)
```
```suggestion
  for (size_t l = 0; l < network.size(); ++l)
```
Tiny style fix. :)

rcurtin commented on src/mlpack/methods/ann/util/check_input_shape.hpp:
```diff
@@ -0,0 +1,53 @@
+    }
+    else
+    {
+      std::string estr = functionName + ": "; 
+                  estr += "the first layer of the network expects ";
+                  estr += std::to_string(layerInShape) + " elements, ";
+                  estr += "but the input has " + std::to_string(inputShape) + " dimensions! ";
```
```suggestion    
      std::string estr = functionName + ": the first layer of the network " +
          "expects " + std::to_string(layerInShape) + " elements, but the " +
          "input has " + std::to_string(inputShape) + " dimensions!";
```
Why not do it all in one line?  Just a simplification suggestion. :)

rcurtin commented on src/mlpack/tests/recurrent_network_test.cpp:
```diff
@@ -868,3 +868,67 @@ TEST_CASE("LargeRhoValueRnnTest", "[RecurrentNetworkTest]")
+  model.Add<Linear<> >(4, 10);
+  model.Add<LogSoftMax<> >();
+
+  std::string expectedMsg = "RNN<>::Train(): ";
+              expectedMsg += "the first layer of the network expects ";
+              expectedMsg += std::to_string(3) + " elements, ";
+              expectedMsg += "but the input has " + std::to_string(1) + " dimensions! ";
```
Personally I think it would be fine just to match the exception type here; we might change the internal output format over time, and it would be tedious to have to update these tests every time. :+1:

(This applies to the FFN test too.)

NippunSharma replied:
I did this because I thought it will be better to add an extra layer of verification that also tests which message is thrown but I see how tedious updating this would be every time message is changed. I will change it 👍

rcurtin commented on src/mlpack/methods/ann/layer/bilinear_interpolation.hpp:
```diff
@@ -118,6 +118,12 @@ class BilinearInterpolation
+  size_t WeightSize() const
+  {
+    return InRowSize;
+  }
```
Should this be `InputShape()`?  Also I think this won't compile; maybe it should be `InRowSize()` or just `inRowSize`?

NippunSharma replied:
You are correct. It should be `InputShape()` and `inRowSize`

rcurtin reviewed (commented):
Awesome work @NippunSharma!  Are you sure that you got all the layer types implemented?  Or are there more where we should open a separate issue to make sure that we handle those?

I have a couple simple suggestions to improve the code---hopefully they're helpful. :+1:

NippunSharma commented:
@rcurtin I have added some more layer types and I think I have covered almost all of the layer types. Some of the layer types which I am having difficulty in doing are : `concat`, `concat_performance`, `minibatch_discrimination`,  `recurrent_attention`.
Are there any other layers apart from these and what I have already done that I am missing ? please tell, if time permits I will try to add them also or open an issue for the same.

shrit commented:
@NippunSharma There is a merge conflict, can you resolve it so we can merge this?

NippunSharma commented:
@shrit , I have resolved the conflicts now

shrit commented:
Perfect I will until all check are green and merge it tomorrow or later this day
